keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1083 @@
+linux-specific settings (such as seccomp profile, run as non root, and disallow privilege
+escalation). If the Kubelet and/or container runtime choose to ignore these linux-specific values at
+runtime, then windows pods should still be allowed under the restricted profile, although the
+profile will not add additional enforcement over baseline (for Windows).
+
+Windows support will be reevaluated prior to this policy feature going to beta, or if/when
+Kubernetes adds support to definitively distinguish between Windows and Linux workloads. See
+[Windows restricted profile support](#windows-restricted-profile-support) for more details.
+
+### Flexible Extension Support
+
+In order to make it as easy as possible to extend and customize this policy controller, we will
+publish the following tools:
+
+- Library implementation - The admission plugin will be written using reusable library code.
+- Webhook implementation - A standalone webhook implementation will be provided (using the same
+  library code), enabling policy enforcement on older clusters.
+- Thorough testing resources, which can be used to validate 3rd party implementations of the policy
+  spec.
+
+### Test Plan
+
+The admission controller can safely be enabled as a no-op with the default-defaults, i.e. everything
+is privileged. This will let us run the admission controller in our standard E2E test jobs, by
+relabeling specific test namespaces.
+
+**E2E Tests:** The following tests should be added:
+
+1. Enforce mode tests:
+    - Test all profile levels
+    - Test profile version support
+2. Warning mode tests:
+    - Profile levels & version support
+3. Namespace policy relabeling
+    - Ensure labeling completes even when there are warnings
+    - Test warning on violating pods
+    - Test dry-run mode
+
+Additionally, we should add tests to the upgrade test suite to ensure that version skew is properly
+handled:
+
+- A minimally specified pod (just a container image) should always be allowed by the baseline
+  policy.
+- A privileged pod should never be allowed by baseline or restricted
+- A Fully specified pod within the bounds of baseline should be allowed by baseline, and rejected by
+  restricted.
+- A minimally specified restricted pod should be allowed at a pinned version.
+
+**Integration Tests:** Audit mode tests should be added to integration testing, where we have
+existing audit logging tests.
+
+**Manual Testing Resources:** Pod resources will be provided covering all dimensions of the baseline
+& restricted profiles, for validation of 3rd party policy implementations. These have been drafted
+by @JimBugwadia: https://github.com/JimBugwadia/pod-security-tests
+
+**Unit Tests:** Both the library and admission controller implementations will have thorough
+coverage of unit tests.
+
+### Monitoring
+
+A single metric will be added to track policy evaluations against pods and [templated pods].
```

Comment:
I'm not sure I understand how the metric is recorded for templated pods. Is there any indication in the metric the evaluation was on a pod or not?

Reply:
That was the resource label, which I thought we decided to nix? Or should we keep it as a simple `{pod, template}` label?

Reply:
I didn't realize we still intended to emit metrics for evaluations against pod templates, since those are non-authoritative (both `enforce` and `warn` just emit warnings but do not block). If we think we'll emit metrics for those, we should distinguish with a label like `resource {pods, ...something other than pods...}`